tests/unit/processing/test_spatial_assets_utils.py

```diff
@@ -0,0 +1,315 @@
+    """
+    Test case for verifying the required metadata is present
+    """
+    spatial_processor = setup_spatial_processor["spatial_processor"]
+    valid_spatial_data = setup_spatial_processor["valid_spatial_data"]
+    library_id = setup_spatial_processor["library_id"]
+
+    # Expected output based on the input data
+    expected_output = {
+        library_id: {
+            "images": {"hires": valid_spatial_data["images"]["hires"], "fullres": []},
+            "scalefactors": {
+                "spot_diameter_fullres": valid_spatial_data["scalefactors"]["spot_diameter_fullres"],
+                "tissue_hires_scalef": valid_spatial_data["scalefactors"]["tissue_hires_scalef"],
+            },
+        }
+    }
+
+    result = spatial_processor.filter_spatial_data(valid_spatial_data, library_id)
+
+    assert result == expected_output, f"Expected {expected_output}, but got {result}"
+
+
+def test__invalid_input_metadata_copy(setup_spatial_processor):
+    """
+    Test case to verify that a KeyError is raised when passing input metadata
+    """
+    spatial_processor = setup_spatial_processor["spatial_processor"]
+    invalid_uns = setup_spatial_processor["invalid_uns"]
+    library_id = setup_spatial_processor["library_id"]
+    with pytest.raises(KeyError):
+        spatial_processor.filter_spatial_data(invalid_uns, library_id)
+
+
+def test__full_high_res_image_present(setup_spatial_processor):
+    """
+    Test that the default image is fullres if present otherwise fall back to hires
+    """
+    spatial_processor = setup_spatial_processor["spatial_processor"]
+    valid_spatial_data = setup_spatial_processor["valid_spatial_data"]
+
+    image_array_uint8 = spatial_processor._prepare_image(valid_spatial_data)
+    assert image_array_uint8.shape == (20, 20, 3)
+
+    valid_hires_uns = copy.deepcopy(valid_spatial_data)
+    valid_hires_uns["images"].pop("fullres", None)
+    image_array_uint8 = spatial_processor._prepare_image(valid_hires_uns)
+    assert image_array_uint8.shape == (10, 10, 3)
+
+    non_valid_uns = copy.deepcopy(valid_hires_uns)
+    non_valid_uns["images"].pop("hires", None)
+    with pytest.raises(KeyError):
+        image_array_uint8 = spatial_processor._prepare_image(non_valid_uns)
+
+
+def test__process_and_flip_image(setup_spatial_processor):
+    """
+    Test the image processing method
+    """
+    spatial_processor = setup_spatial_processor["spatial_processor"]
+    test_image_array = (np.random.rand(10, 10, 3) * 255).astype(np.uint8)
+    expected_flipped_array = np.flipud(test_image_array)
+
+    flipped_image_array = spatial_processor._process_and_flip_image(test_image_array)
+    assert np.array_equal(flipped_image_array, expected_flipped_array)
+
+
+def test__crop_to_aspect_ratio(setup_spatial_processor):
+    """
+    Test the cropping method to ensure the image is cropped to a square
+    """
+    spatial_processor = setup_spatial_processor["spatial_processor"]
+    width, height = 23, 11
```

Comment:
Does this also work with even numbers?

Reply:
you can use [@pytest.mark.parameterize](https://docs.pytest.org/en/7.4.x/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions) to test that both work using the same test.

Reply:
added!